modules/post/windows/gather/enum_onedrive.rb

```diff
@@ -0,0 +1,176 @@
+          across all users (providing access is permitted). It is useful for identifying document libraries
+          that may otherwise not be obvious which could contain sensitive or useful information.
+        },
+        'License' => MSF_LICENSE,
+        'Platform' => ['win'],
+        'SessionTypes' => ['meterpreter'],
+        'Author' => ['Stuart Morgan <stuart.morgan[at]mwrinfosecurity.com>']
+      )
+    )
+  end
+
+  def display_report(sid, info, sync_used, sync_all, results_table)
+    info.each do |key, result|
+      next if result['ScopeIdToMountPointPathCache'].nil? || result['ScopeIdToMountPointPathCache'].empty?
+
+      row = []
+      print_line
+      print_line "  #{key}"
+      print_line "  #{'=' * key.length}"
+      print_line
+      row << sid
+      row << key
+      ONEDRIVE_ACCOUNT_KEYS.each do |col|
+        row << result[col].to_s
+        if result['Business'] == '1' || PERSONAL_ONEDRIVE_KEYS.include?(col)
+          print_line "    #{col}: #{result[col]}"
+        end
+      end
+      result['ScopeIdToMountPointPathCache'].each do |scopes|
+        subrow = row.clone
+        print_line
+        SYNC_ENGINES_KEYS.each do |sync|
+          subrow << scopes[sync].to_s
+          print_line "    | #{sync}: #{scopes[sync]}"
+        end
+        results_table << subrow
+      end
+    end
+
+    sync_all_list = []
+    sync_all.each do |key, _result|
+      sync_all_list.push(key)
+    end
+
+    diff = sync_all_list - sync_used
+    if !(diff.nil? || diff.empty?)
+      print_line
+      print_line '  ORPHANED'
+      print_line '  ========'
+      diff.each do |scopeid|
+        csvrow = []
+        print_line
+        # Augment the CSV
+        csvrow << sid
+        csvrow << ''
+        ONEDRIVE_ACCOUNT_KEYS.each do |_od|
+          csvrow << ''
+        end
+        SYNC_ENGINES_KEYS.each do |sync|
+          csvrow << sync_all[scopeid][sync]
+          print_line "  #{sync}: #{sync_all[scopeid][sync]}"
+        end
+        results_table << csvrow
+      end
+    end
+  end
+
+  def get_syncengine_data(master, syncengines)
+    all_syncengines = {}
+    syncengines.each do |ses|
+      newses = {}
+      SYNC_ENGINES_KEYS.each do |key|
+        newses[key] = registry_getvaldata("#{master}\\#{ses}", key).to_s
+      end
+      all_syncengines[ses] = newses
+    end
+    all_syncengines
+  end
+
+  def get_onedrive_accounts(reg, accounts, syncdata)
+    all_oda = {}
+    synctargets_used = []
+    ret = {}
+    reg.each do |ses|
+      newses = {}
+      ONEDRIVE_ACCOUNT_KEYS.each do |key|
+        newses[key] = registry_getvaldata("#{accounts}\\#{ses}", key).to_s
+      end
+      scopeids = registry_enumvals("#{accounts}\\#{ses}\\ScopeIdToMountPointPathCache")
+      next if scopeids.nil? || scopeids.empty?
+
+      newses['ScopeIdToMountPointPathCache'] = []
+      scopeids.each do |sid|
+        target = syncdata[sid]
```

Comment:
Correct me if I am wrong but there seems to be the potential for an invalid offset access here. The reason for this is that the value of `sid` is indirectly obtained by calling `registry_enumvals` and then is directly used as an array entry within the `syncdata` array without first checking if `sid` is a valid entry within the `syncdata` array.

This could cause issues if you query the registry and the registry data, which shouldn't be trusted, is corrupted somehow.

This is actually even more important given the line `newses['ScopeIdToMountPointPathCache'].push(target)` on line 124.

Reply:
Gah so technically speaking the checks should be done in `get_onedrive_accounts()` rather than in `display_report` which should be purely responsible for displaying the report rather than filtering out the contents of the results as well. However I do see that we have the following line in `display_report()` which prevents this from being an issue:

```
next if result['ScopeIdToMountPointPathCache'].nil? || result['ScopeIdToMountPointPathCache'].empty?
```

So whilst I don't agree with the approach taken here r.e separating the display logic from the parsing logic, this is still technically a valid approach and this issue should be marked as resolved.

Reply:
Cheers, I'll bear that in mind in future. I accept that there isn't clear separation here.